### ISSUE TITLE
java.lang.IllegalStateException: Objects cannot be changed outside a transaction on findAll() call

### ISSUE THREAD

**Issue by mmvicsonvictor** [realm/realm-java#941]
I am getting this IllegalStateException on the findAll() call. I have a android library which contains the RealmInstanceFactory class, which basically produces realm instances(no caching). The library is in the 'aars' directory of the app that I am trying to test with.  

I have no clue why I am getting this error. Can you please help?

    public List<AppSetupHistory> getAllAppSetupHistoryForType(String type){
        Realm realm = realmInstanceFactory.getInstance();
        List<RealmAppSetupHistory> historyEntries = realm.where(RealmAppSetupHistory.class)
                .equalTo("type", type).findAll();
        List<AppSetupHistory> result = new ArrayList<AppSetupHistory>();
        for (RealmAppSetupHistory history : historyEntries){
            result.add(new AppSetupHistory(history));
        }
        realm.close();
        Log.i(LOG_TAG, "No. of AppSetupHistory entries found for type: " + result.size());
        //Log.i(LOG_TAG, "No. of AppSetupHistory entries found for type: " + result.size());
        return result;
    }    

Caused by: java.lang.IllegalStateException: Objects cannot be changed outside a transaction; see beginTransaction() for details.
            at io.realm.internal.Group.throwImmutable(Group.java:320)
            at io.realm.internal.Group.getTable(Group.java:220)
            at io.realm.Realm.getTable(Realm.java:323)
            at io.realm.RealmQuery.<init>(RealmQuery.java:75)
            at io.realm.Realm.where(Realm.java:1279)
            at com.mm.androidshared.db.realm.dao.RealmDao.getAllAppSetupHistoryForType(RealmDao.java:232)
            at com.customername.androidui.builder.AppSetupBuilder.getHistoryForType(AppSetupBuilder.java:56)
            at com.customername.androidui.builder.AppSetupBuilder.execute(AppSetupBuilder.java:72)

**Comment by mmvicsonvictor** [comment]
BTW, I am using the 0.79.1 realm jar, which is packaged in the android archive that I am using in the app.

**Comment by mmvicsonvictor** [comment]
Digging more, found some other issue referring to the "ValidationList". So the Android Library project that I am using contains a ValidationList for the sample classes(extending RealmObject) in the library itself. However, I have a different set of RealmObject classes in my actual app module and I do not see a ValidationList class generated for it. I do however see the Proxy classes created for them.

**Comment by cmelchior** [comment]
Hi @mmvicsonvictor 
Yes, currently Realm doesn't work if you model classes is split between a library project and the app project itself. This is due to how the annotation processor work, as it doesn't process already compiled code (ie. your library).

It is possible to get around it by manually adding the files to the ValidationList files, but it is a quite error prone and manual process.

But we are working on a new approach for use cases like that allow you to share these classes between projects. You can follow progress here: https://github.com/realm/realm-java/pull/938

**Comment by mmvicsonvictor** [comment]
@cmelchior  Thanks for the quick response. What I am trying to accomplish is to have a realm library archive(aar), which can be versioned and maintained separately. Designed to be pretty light, it would just offer realm bootstrapping with some examples(RealmObject types and a DAO to show how to use it) and junit support(ofcourse when roboelectric support happens on realm). So when the Realm annotation processor runs on this library, it will produce the Proxies and the ValidationList.

And this aar would be used in the app modules, which themselves will have their own set of RealmObjects, DAOs etc. And I need the Annotation Processor to run on this module as well. And either produce a consolidated validation list(or produce an additional one for the RealmObjects in the app module) and produce proxies as well.

Would the updated Annotation Processor work this way? And is it available in the master branch for me to try? Also, is it coming out in the next version. 

I am so close to dumping DB4O, but yet so far. Null support, robolectric support are the other key things which is still stopping me. :)

Thanks.

**Comment by cmelchior** [comment]
The PR I am referring to is not in master yet, as we havn't decided on the API yet, so a release date is still uncertain, but it will produce something called a RealmClassCollection pr. project that you can combine to make a Realm aware of all classes. So it would fit your scenario perfectly.

**Comment by mmvicsonvictor** [comment]
So a "RealmClassCollection" which will get generated per module(along with the proxies I assume). And would the process of combining them be provided by the library(a manual step would be error prone and leave room for more issues)?

**Comment by cmelchior** [comment]
The current proposal require the library developer to manually create the RealmClassCollection, as you 
might not want to expose all model objects, like so:

```
@RealmModule
public MyModule extends RealmClassCollection {
  public Set<Class<? extends RealmObject>> getModuleClasses() {
    return new HashSet<>(Arrays.asList(Foo.class, Bar.class)); 
  }
}

// And the app developer would so something like:
Realm.addModule(new MyModule())
```

However it involves some work from the library developer especially maintaining the classes. There is no around the library developer having to create a public class for the app code (due to the annotation processor), but adding some way to automatically include all model classes might be a good idea, something like

```
@RealmModule(includeAll = true)
public MyModule extends RealmClassCollection {
  public Set<Class<? extends RealmObject>> getModuleClasses() {
    return new HashSet<>(Arrays.asList(Foo.class, Bar.class)); 
  }
}
```

**Comment by mmvicsonvictor** [comment]
Makes sense. Just curious, where would the following code go? Currently there is not bootstrapping that is required in the application code.

Realm.addModule(new MyModule()) 

Also, how would the proxy generation work across modules?

**Comment by cmelchior** [comment]
It is part of a bigger rewrite as part of adding a new migration API.

We are considering adding a RealmConfiguration to control how Realms are created, this will also include adding additional modules:
https://github.com/realm/realm-java/pull/929

And how the proxies interact across modules has also been rewritten to support both library projects and ProGuard better:
https://github.com/realm/realm-java/pull/938

See this JavaDoc for a more detailed description on the internals (still work-in-progress so a bit rough around the edges):
https://github.com/realm/realm-java/blob/cm-realm-modules/realm-annotations-processor/src/main/java/io/realm/processor/RealmProcessor.java

**Comment by mmvicsonvictor** [comment]
Thanks for sharing the info. Really looking forward to this work getting to the main branch.

**Comment by mmvicsonvictor** [comment]
@cmelchior  I was looking at the javadoc and also the code a bit and I have a question. Looks like we cannot have normal and library modules in the same project(I am assuming you are referring to an Android project in AS or the like).

        if (modules.size() > 0 && libraryModules.size() > 0) {
            error("Normal modules and library modules cannot be mixed in the same project");
            return true;
        }

But that is what exactly libraries would be, is it not? We might have 1 or n libraries(aar) where the modules are defined as a "library". And one app module which imports all these libraries and in addition to it defines its own module, which is NOT a library module.

Did I read that wrong?

**Comment by cmelchior** [comment]
Actually the modules there referer to RealmModules, no AndroidStudio modules, but the distinction is pretty much the same. 

What is important is that Library projects are "special" and require the annotation processor to do something slightly different. The annotation processor cannot tell the difference automatically and needs help through the annotations (Hence @RealmModule(library = true)). Mixing library and normal RealmModules means the annotation processor cannot decide which type of project it is and throws an error.

The reason this distinction is important is that "app" projects automatically generate a "DefaultRealmModule", If all library projects did this as well we would have name conflicts.

When actually running the app. You will be able to freely mix the generated modules any way you like, so yes you are reading it wrong, but that is probably just a sign that the documentation / error messages need improvement :)





Conveying that distinction is a bit tricky though, which is probably also why we are going to iterate on the naming / API a bit before settling on the specific naming.

**Comment by mmvicsonvictor** [comment]
So if I hear you right, I can have n number of android archives/libraries - referenced in my app - which have 1 or more Realm modules in them(all of them are library modules). And I can have another RealmModule defined in the app source which is NOT a library module. Correct?

**Comment by cmelchior** [comment]
Yes that was the idea.
That way you combine exactly how you schema will look for any given Realm no matter where the model class came from. This can eg. also be useful for just app projects if you want one Realm only holding Preferences and another holding User data. Separating them into modules make it easier to think about versioning and migrations

**Comment by mmvicsonvictor** [comment]
Exactly! Thanks!

**Comment by mmvicsonvictor** [comment]
@cmelchior As a temp solution, I was trying the following hack. I tried to not run the AnnotationProcessor on the library project by excluding the RealmClass objects in the source set. And it did create the proxies and the validation list and also the RealmJsonImpl class.

But when I assemble and deploy the app which uses the library, it looks like the Realm annotation processor did not run. Do you know if I can hack this way to get it produce the ValidationList, proxies and the JsonImpl for the classes in the main app?

     Caused by: io.realm.exceptions.RealmException: Could not find io.realm.RealmJsonImpl
            at io.realm.Realm.getRealmJson(Realm.java:263)
            at io.realm.Realm.<init>(Realm.java:161)
            at io.realm.Realm.createAndValidate(Realm.java:516)
            at io.realm.Realm.create(Realm.java:482)
            at io.realm.Realm.getInstance(Realm.java:397)

**Comment by cmelchior** [comment]
Hi @mmvicsonvictor 
Completely disabling the annotation processor is probably a bad idea. Instead the hacky approach would be to take the RealmJsonImpl and ValidationList from you library project and merge the code manually with the code generated by the app project. It doesn't scale very well and is a highly manual process, so I would not really recommend it.

**Comment by mmvicsonvictor** [comment]
@cmelchior Since either approaches are not so ideal. I changed my plans for the time being. I pulled out the models and the unit tests out of the library. So all it has at this point is Realm bootstrapping that we want to use as some patterns. And the application module has the .aar of the library and has some Android tests to test out the DAOs.

I would really love to see this issue resolved, along with the support for roboelectric and high level APIs for migration. Oh and the support for NULL values :) As it stands right now, there is no way to test Realm dependencies on a JVM and thats a showstopper for us.

Also, it looks like we cannot have AndroidTests in a android library module, which sucks. This is not a Realm issue however.
